controllers/serviceclaim_controller.go

```diff
@@ -47,13 +61,229 @@ type ServiceClaimReconciler struct {
+		l.Info("unable to retrieve ServiceClaim", "error", err)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
 
+	var rsl primazaiov1alpha1.RegisteredServiceList
+	lo := client.ListOptions{Namespace: req.NamespacedName.Namespace}
+	if err := r.List(ctx, &rsl, &lo); err != nil {
+		l.Info("unable to retrieve RegisteredServiceList", "error", err)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	rsName, err := r.processServiceClaim(ctx, req, rsl, sclaim)
+	if err != nil {
+		l.Error(err, "unable while processing ServiceClaim")
+		return ctrl.Result{}, err
+	}
+	sclaim.Status.State = "Resolved"
+	sclaim.Status.ClaimID = uuid.New().String()
+	sclaim.Status.RegisteredService = rsName
+	sclaim.Status.ServiceBinding = req.NamespacedName.Name
+	if err := r.Status().Update(ctx, &sclaim); err != nil {
+		l.Error(err, "unable to update the ServiceClaim", "ServiceClaim", sclaim)
+		return ctrl.Result{}, err
+	}
 	return ctrl.Result{}, nil
 }
 
+// Ref. https://stackoverflow.com/a/18879994/547840
+func checkSCISubset(serviceClaim, registeredService []v1alpha1.ServiceClassIdentityItem) bool {
+	set := make(map[v1alpha1.ServiceClassIdentityItem]int)
+	for _, value := range registeredService {
+		set[value] += 1
+	}
+
+	for _, value := range serviceClaim {
+		if count, found := set[value]; !found {
+			return false
+		} else if count < 1 {
+			return false
+		} else {
+			set[value] = count - 1
+		}
+	}
+
+	return true
+}
+
+func (r *ServiceClaimReconciler) processServiceClaim(
+	ctx context.Context,
+	req ctrl.Request,
+	rsl primazaiov1alpha1.RegisteredServiceList,
+	sclaim primazaiov1alpha1.ServiceClaim,
+) (string, error) {
+	l := log.FromContext(ctx)
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.NamespacedName.Name,
+			Namespace: req.NamespacedName.Namespace,
+		},
+		StringData: map[string]string{},
+	}
+	var rsName string
+	// count the number of secret data entries
+	count := 0
+	// loop over every RegisteredService
+	for _, rs := range rsl.Items {
+		// Check if the ServiceClassIdentity given in ServiceClaim is a subset of
+		// ServiceClassIdentity given in the RegisteredService
+		if checkSCISubset(sclaim.Spec.ServiceClassIdentity, rs.Spec.ServiceClassIdentity) {
+			// loop over the environments given in the constraints part of the RegisteredService
+			for _, e := range rs.Spec.Constraints.Environments {
+				// check if the EnvironmentTag of ServiceClaim matches tat of the environment constraint
+				if sclaim.Spec.EnvironmentTag == e {
+					// loop over the ServiceEndpointDefinition array part of RegisteredService
+					for _, sed := range rs.Spec.ServiceEndpointDefinition {
+						// check if the value is non-empty
+						if sed.Value != "" {
+							// check if the ServiceEndpointDefinitionKeys part of ServiceClaim has the current
+							// SED name in the RegisteredService
+							if itemContains(sclaim.Spec.ServiceEndpointDefinitionKeys, sed.Name) {
```

Comment:
I think this loop deserves its own function. Also the we should try to reduce the nesting here.
A simple approach may be to break that down to multiple functions and change `if-else`s in the following way:

```
if condition {
  do_true()
else {
  do_false()
}
```
to
```
if !condition {
   do_false()
   return
}

do_true()
```

Reply:
Done